### ISSUE TITLE
Question about the cosine distance.

### ISSUE THREAD

**Issue by z1358** [Sha-Lab/FEAT#62]
Thanks for your work！ There is a question about the cosine distance.        
`if self.args.use_euclidean:
            query = query.view(-1, emb_dim).unsqueeze(1) # (Nbatch*Nq*Nw, 1, d)
            proto = proto.unsqueeze(1).expand(num_batch, num_query, num_proto, emb_dim).contiguous()
            proto = proto.view(num_batch*num_query, num_proto, emb_dim) # (Nbatch x Nq, Nk, d)

            logits = - torch.sum((proto - query) ** 2, 2) / self.args.temperature
        else:
            proto = F.normalize(proto, dim=-1) # normalize for cosine distance
            query = query.view(num_batch, -1, emb_dim) # (Nbatch,  Nq*Nw, d)

            logits = torch.bmm(query, proto.permute([0,2,1])) / self.args.temperature
            logits = logits.view(-1, num_proto)`
When the argument is false, i.e. when using cosine distance, the query should also be normalized, right?

**Comment by Han-Jia** [comment]
Hi, this is a good question. 

The usual cosine distance normalizes two inputs simultaneously. When we make predictions following the ProtoNet, we compute the distance between one instance with multiple (N) prototypes and find the maximum one.  Since all the cosine values divide the same norm of the instance, so only normalizing the prototypes has the same effect as normalizing both of the prototypes and the instances.

In addition, we find using such a semi-normalized cosine distance can get better performance than the traditional cosine distance, which does not need careful tunning on the temperature. Please see our [recent paper](https://arxiv.org/abs/2011.14663), Section 4.1 for more details.

**Comment by yohann84L** [comment]
Hey @Han-Jia , about your paper `Revisiting Unsupervised Meta-Learning: Amplifying or Compensating for the Characteristics of Few-Shot Tasks`, did you release the code ?

**Comment by Han-Jia** [comment]
Hi,

We will release the code at https://github.com/hanlu-nju/revisiting-UML later this month.
